### PR TITLE
Fix bug DynamicSupervisor child_spec validation

### PR DESCRIPTION
**Pull request by eksperimental**
It would not accept 0 as a valid shutdown value.
It only happened in DynamicSupervisor, but tests are added for both DynamicSupervisor and Supervisor modules.

### CODE REVIEW

**Comment by eksperimental** [comment]
This commit can also be applied to the v1.13 branch

**Comment by josevalim** [comment]
I believe shutdown >= 0 is recent in OTP 24. So you will need a conditional in supervisor tests.

**Comment by eksperimental** [comment]
> I believe shutdown >= 0 is recent in OTP 24. So you will need a conditional in supervisor tests.

Yes, I was seeing that in the CI.

**Comment by eksperimental** [comment]
I am wondering if the validation should also consider the OTP version.

**Comment by eksperimental** [comment]
... or update the specs so shutdown is `pos_integer()` and when OTP is the minimum required version, apply this PR

**Comment by josevalim** [comment]
Waiting until OTP 24 is the easiest route for sure.

**Comment by eksperimental** [comment]
What about the spec?

**Comment by josevalim** [comment]
We should fix it to mirror whatever we support today, yeah.

**Comment by eksperimental** [comment]
Ok, I will fix and leave a TODO when we support OTP 24+ exclusively including a link to this PR.

File: lib/elixir/test/elixir/dynamic_supervisor_test.exs [hunk 1]
```diff
@@ -451,6 +458,33 @@ defmodule DynamicSupervisorTest do
       assert_receive {:EXIT, ^pid, :shutdown}
     end
 
+    # == 
+
```
Comment (josevalim) [inline lib/elixir/test/elixir/dynamic_supervisor_test.exs:462]:
```suggestion
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**